programs/openbook-v2/src/state/orderbook/queue.rs

```diff
@@ -39,117 +49,165 @@ impl EventQueue {
         self.len() == 0
     }
 
-    pub fn full(&self) -> bool {
-        self.header.count() == self.buf.len()
+    pub fn is_full(&self) -> bool {
+        self.len() == self.nodes.len()
     }
 
-    pub fn push_back(&mut self, value: AnyEvent) -> std::result::Result<(), AnyEvent> {
-        if self.full() {
-            return Err(value);
+    pub fn front(&self) -> Option<&AnyEvent> {
+        if self.is_empty() {
+            None
+        } else {
+            Some(&self.nodes[self.header.used_head()].event)
         }
-        let slot = (self.header.head() + self.header.count()) % self.buf.len();
-        self.buf[slot] = value;
-
-        let count = self.header.count();
-        self.header.set_count((count + 1) as u32); // guaranteed because of full() check
-
-        self.header.incr_event_id();
-        Ok(())
     }
 
-    pub fn peek_front(&self) -> Option<&AnyEvent> {
-        if self.is_empty() {
-            return None;
+    pub fn at_slot(&self, slot: usize) -> Option<&AnyEvent> {
+        if self.nodes[slot].is_free() {
+            None
+        } else {
+            Some(&self.nodes[slot].event)
         }
-        Some(&self.buf[self.header.head()])
     }
 
-    pub fn peek_front_mut(&mut self) -> Option<&mut AnyEvent> {
+    pub fn push_back(&mut self, value: AnyEvent) {
+        assert!(!self.is_full());
+
+        let slot = self.header.free_head;
+        self.header.free_head = self.nodes[slot as usize].next;
+
+        let new_next: u16;
+        let new_prev: u16;
+
         if self.is_empty() {
-            return None;
+            new_next = slot;
+            new_prev = slot;
+
+            self.header.used_head = slot;
+        } else {
+            new_next = self.header.used_head;
+            new_prev = self.nodes[new_next as usize].prev;
+
+            self.nodes[new_prev as usize].next = slot;
+            self.nodes[new_next as usize].prev = slot;
         }
-        Some(&mut self.buf[self.header.head()])
+
+        self.header.incr_count();
+        self.header.incr_event_id();
+        self.nodes[slot as usize].event = value;
+        self.nodes[slot as usize].next = new_next;
+        self.nodes[slot as usize].prev = new_prev;
     }
 
     pub fn pop_front(&mut self) -> Result<AnyEvent> {
-        require!(!self.is_empty(), OpenBookError::SomeError);
+        self.delete_slot(self.header.used_head())
+    }
 
-        let value = self.buf[self.header.head()];
+    pub fn delete_slot(&mut self, slot: usize) -> Result<AnyEvent> {
+        if self.is_empty() || self.nodes[slot].is_free() {
```

Comment:
this is going to panic if `slot > nodes.len()`. Might be nicer to return an error?